bridge-ffi/src/lib.rs

```diff
@@ -0,0 +1,833 @@
+            }
+        }
+    });
+}
+
+/// Shutdown registered worker.
+///
+/// The req_proto and req_proto_len represent a byte array for a
+/// ShutdownWorkerRequest protobuf message. The callback is invoked on
+/// completion with a ShutdownWorkerResponse protobuf message.
+#[no_mangle]
+pub extern "C" fn tmprl_shutdown_worker(
+    core: *mut tmprl_core_t,
+    #[allow(unused_variables)] // We intentionally ignore the request
+    req_proto: *const u8,
+    #[allow(unused_variables)] req_proto_len: libc::size_t,
+    user_data: *mut libc::c_void,
+    callback: tmprl_callback,
+) {
+    let core = unsafe { &mut *core };
+    let req =
+        match tmprl_core_t::decode_proto::<bridge::ShutdownWorkerRequest>(req_proto, req_proto_len)
+        {
+            Ok(req) => req,
+            Err(message) => {
+                let resp = bridge::ShutdownWorkerResponse {
+                    error: Some(bridge::shutdown_worker_response::Error { message }),
+                };
+                unsafe {
+                    callback(user_data, core.encode_proto(&resp).into_raw());
+                }
+                return;
+            }
+        };
+    let user_data = UserDataHandle(user_data);
+    core.tokio_runtime.clone().spawn(async move {
+        core.shutdown_worker(req).await;
+        unsafe {
+            callback(user_data.into(), &*DEFAULT_SHUTDOWN_WORKER_RESPONSE_BYTES);
+        }
+    });
+}
+
+/// Poll workflow activation.
+///
+/// The req_proto and req_proto_len represent a byte array for a
+/// PollWorkflowActivationRequest protobuf message. The callback is invoked on
+/// completion with a PollWorkflowActivationResponse protobuf message.
+#[no_mangle]
+pub extern "C" fn tmprl_poll_workflow_activation(
+    core: *mut tmprl_core_t,
+    req_proto: *const u8,
+    req_proto_len: libc::size_t,
+    user_data: *mut libc::c_void,
+    callback: tmprl_callback,
+) {
+    let core = unsafe { &mut *core };
+    let req = match tmprl_core_t::decode_proto::<bridge::PollWorkflowActivationRequest>(
+        req_proto,
+        req_proto_len,
+    ) {
+        Ok(req) => req,
+        Err(message) => {
+            let resp = bridge::PollWorkflowActivationResponse {
+                response: Some(bridge::poll_workflow_activation_response::Response::Error(
+                    bridge::poll_workflow_activation_response::Error {
+                        message,
+                        shutdown: false,
+                    },
+                )),
+            };
+            unsafe {
+                callback(user_data, core.encode_proto(&resp).into_raw());
+            }
+            return;
+        }
+    };
+    let user_data = UserDataHandle(user_data);
+    core.tokio_runtime.clone().spawn(async move {
+        let resp = bridge::PollWorkflowActivationResponse {
+            response: Some(match core.poll_workflow_activation(req).await {
+                Ok(act) => bridge::poll_workflow_activation_response::Response::Activation(act),
```

Comment:
Nit: Dunno if we need to prefix everything with `bridge::` rather than just importing directly

Reply:
I can change if you want. This reads better to me, but I don't mind adding a bunch of `use` statements for all the proto types if you want. Of course with this specific case, `Response` is ambiguous, so I'd still need `poll_workflow_activation_response::Response` and import that module, but for things like `bridge::PollWorkflowActivationResponse` I can import the type instead of the module.